lv2: reduce closure

jonringer opened the pull request:
###### Motivation for this change
gtk2 is optional

Before:
```
$ nix path-info -Sh ./result
/nix/store/ysdp0p56qpnx9kd6lyyx6rgbwnz97wq6-lv2-1.18.2	 327.1M
```
After:
```
$ nix path-info -Sh ./result
/nix/store/q1rcs1dkj4zrrd0nd3qk3dnn5akh1b9g-lv2-1.18.2	 141.5M
```

related: https://github.com/NixOS/nixpkgs/issues/159612

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Commit:
lv2: drop GTK 2 dependency

gtk2 is optional and only used for some example plug-ins. In particular,
removing `gtk2` disables building UI module for `$out/lib/lv2/eg-sampler.lv2`
(`<http://lv2plug.in/plugins/eg-sampler#ui>`) and the whole
`$out/lib/lv2/eg-scope.lv2` (`<http://lv2plug.in/plugins/eg-scope>`).

jonringer commented on pkgs/development/libraries/audio/lv2/default.nix:
```diff
@@ -9,8 +9,8 @@ stdenv.mkDerivation rec {
 
-  nativeBuildInputs = [ pkg-config wafHook ];
-  buildInputs = [ gtk2 libsndfile python3 ];
+  nativeBuildInputs = [ pkg-config wafHook python3 ];
```
oh, that's where that's coming from

jonringer replied:
that's really awkward for cross compilation

jtojnar replied:
Maybe we could move the file to a `dev` output? Not really sure about what it is used for.

jonringer replied:
done

jonringer replied:
also, dev output fixes a lot of issues, including the closure propagation

Commit:
lv2: add pipewire test

; Conflicts:
;	pkgs/development/libraries/audio/lv2/default.nix

Commit:
lv2: add dev output

Also put programs there since those are just development tools,
that would bring unnecessary Python dependency into out’s runtime closure.

jtojnar commented on pkgs/development/libraries/audio/lv2/default.nix:
```diff
@@ -1,19 +1,29 @@
   pname = "lv2";
   version = "1.18.2";
 
+  outputs = [ "out" "dev" ];
```
Probably also need `outputBin = "dev";`.

jtojnar commented on pkgs/applications/audio/magnetophonDSP/LazyLimiter/default.nix:
```diff
@@ -10,7 +11,7 @@ stdenv.mkDerivation rec {
   };
 
-  buildInputs = [ faust2jaqt faust2lv2 ];
+  buildInputs = [ faust2jaqt faust2lv2 lv2 ];
```
Right, this should be fixed by correcting the symlinks in `lv2.dev` to point to `lv2.out`.

jonringer commented:
honestly, I lost all will to continue this, this package is just a nebulous web of symlinks and hacks.

jtojnar commented:
Looking at the code, it should work if we pass the correct `includedir` flag to waf:

https://github.com/lv2/lv2/blob/6cefc7df1a6158c79d23029df183c09b10b88cad/wscript#L347-L355

jtojnar commented:
Yeah, the relativization is ugly but seems to work:

```patch
--- a/pkgs/development/libraries/audio/lv2/default.nix
+++ b/pkgs/development/libraries/audio/lv2/default.nix
@@ -18,7 +18,11 @@ stdenv.mkDerivation rec {
   nativeBuildInputs = [ pkg-config wafHook ];
   buildInputs = [ libsndfile python3 ];
 
-  wafConfigureFlags = lib.optionals stdenv.isDarwin [ "--lv2dir=${placeholder "out"}/lib/lv2" ];
+  wafConfigureFlags = [
+    "--includedir=${placeholder "dev"}/include"
+  ] ++ lib.optionals stdenv.isDarwin [
+    "--lv2dir=${placeholder "out"}/lib/lv2"
+  ];
 
   passthru.tests = {
     inherit pipewire;
```

siraben commented:
Result of `nixpkgs-review pr 159875` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages marked as broken and skipped:</summary>
  <ul>
    <li>audacity</li>
    <li>rkrlv2</li>
    <li>suil-qt5</li>
  </ul>
</details>
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li>audacity-gtk2 (audacity-gtk3)</li>
    <li>gxmatcheq-lv2</li>
    <li>gxplugins-lv2</li>
    <li>suil (suil-qt4)</li>
  </ul>
</details>
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>lilv</li>
    <li>lv2</li>
    <li>lvtk</li>
    <li>sratom</li>
  </ul>
</details>

ajs124 commented on pkgs/development/libraries/audio/lv2/default.nix:
```diff
@@ -1,18 +1,32 @@
-{ lib, stdenv, fetchurl, gtk2, libsndfile, pkg-config, python3, wafHook }:
```
If we're dropping `gtk2` from `buildInputs`, shouldn't we also drop it here?

Also, this probably supersedes #151292 then (cc @cheriimoya).

jonringer replied:
I'm not dropping it anymore.

Exclusion of gtk2 also makes generation of some extensions not possible.

jonringer replied:
oh, looks like I forgot to push some changes.. I'll be back on my desktop tomorrow.

jtojnar replied:
I think at this point it is preferable to drop `gtk2`. If someone uses the `scope` example plug-in for tests like [Carla](https://github.com/falkTX/Carla/blob/b8d38e395c215eed189a4849972f90af810ef971/source/tests.old/CarlaUtils2.cpp#L88) did, they can easily replace it with another plug-in. If they really want to use that plug-in, they can build a separate derivation. I would not expect anyone to actually need that and getting rid of GTK 2 on the ISO would be nice.

cheriimoya replied:
> I think at this point it is preferable to drop `gtk2`.

Exactly what i thought when opening #151292

jtojnar replied:
Amended the commit to drop it here as well.

jtojnar committed:
lv2: format the expression
